src/platform/datapath_epoll.c

```diff
@@ -2579,27 +2211,204 @@ CxPlatSocketSend(
     _In_ uint16_t IdealProcessor
     )
 {
+    UNREFERENCED_PARAMETER(Socket);
     UNREFERENCED_PARAMETER(IdealProcessor);
-    QUIC_STATUS Status =
-        CxPlatSocketSendInternal(
-            Socket,
-            &Route->LocalAddress,
-            &Route->RemoteAddress,
-            SendData,
-            FALSE);
+
+    //
+    // Finalize the state of the send data and log the send.
+    //
+    CxPlatSendDataFinalizeSendBuffer(SendData);
+    QuicTraceEvent(
+        DatapathSend,
+        "[data][%p] Send %u bytes in %hhu buffers (segment=%hu) Dst=%!ADDR!, Src=%!ADDR!",
+        Socket,
+        SendData->TotalSize,
+        SendData->BufferCount,
+        SendData->SegmentSize,
+        CASTED_CLOG_BYTEARRAY(sizeof(Route->RemoteAddress), &Route->RemoteAddress),
+        CASTED_CLOG_BYTEARRAY(sizeof(Route->LocalAddress), &Route->LocalAddress));
+
+    //
+    // Cache the address, mapping the remote address as necessary.
+    //
+    CxPlatConvertToMappedV6(&Route->RemoteAddress, &SendData->RemoteAddress);
+    SendData->LocalAddress = Route->LocalAddress;
+
+    //
+    // Check to see if we need to pend because there's already queue.
```

Comment:
To align with Windows user mode, I think we should **always** queue to the datapath thread (which may or may not be different from the QUIC thread) to do the TX, otherwise I don't think we actually get much (any?) parallelization of the QUIC and UDP logic on TX.